File: src/sarma_cli/engine/agent_runner.py
"""Agent construction and streaming execution for Ruflo and audit turns.

For audit-mode runs, uses LangGraph subgraph streaming
(``subgraphs=True, version="v2"``) so that every token / tool call /
result is attributable to either the orchestrator or a named subagent.
The :class:`EventTranslator` resolves namespace tuples into subagent
names and emits :class:`StreamEvent` instances with a ``subagent``
field on every payload.

For Ruflo, uses a primary LangGraph ReAct agent with controlled delegation.
"""

from __future__ import annotations

import logging
from typing import Any, AsyncIterator

from sarma_cli.engine.agent_factory import AgentFactory
from sarma_cli.engine.dto import McpServerDTO, ModelProviderDTO
from sarma_cli.engine.errors import AgentRunError
from sarma_cli.engine.mcp_pool import McpClientPool
from sarma_cli.engine.models import (
    AgentRunConfig,
    ConversationMessage,
    ResolvedSkill,
    StreamEvent,
)
from sarma_cli.engine.streaming import EventTranslator
from sarma_cli.engine.enums import StreamEventType

logger = logging.getLogger(__name__)


class AgentRunner:
    """Build a LangGraph agent, run it with streaming, and yield events."""

    def __init__(
        self,
        factory: AgentFactory,
        pool: McpClientPool,
        provider: ModelProviderDTO,
        enabled_servers: list[McpServerDTO],
        skill: ResolvedSkill | None,
        history: list[ConversationMessage],
        system_prompt: str,
        conversation_id: str,
        turn_id: str,
        mode: str = "audit",
        subagent_providers: dict[str, ModelProviderDTO] | None = None,
        subagent_mcp_allow: dict[str, list[str] | None] | None = None,
        subagent_skills: dict[str, ResolvedSkill | None] | None = None,
    ) -> None:
        self._factory = factory
        self._pool = pool
        self._provider = provider
        self._enabled_servers = enabled_servers
        self._skill = skill
        self._history = history
        self._system_prompt = system_prompt
        self._conversation_id = conversation_id
        self._turn_id = turn_id
        self._mode = mode
        self._subagent_providers = subagent_providers or {}
        self._subagent_mcp_allow = subagent_mcp_allow or {}
        self._subagent_skills = subagent_skills or {}
        self.assistant_content = ""
        self.reasoning_content = ""
        self.tool_calls: list[StreamEvent] = []
        self.run_config: AgentRunConfig | None = None

    async def run(self, message: str) -> AsyncIterator[StreamEvent]:
        self.run_config = AgentRunConfig(
            conversation_id=self._conversation_id,
            provider=self._provider,
            skill=self._skill,
            enabled_servers=self._enabled_servers,
            message_history=self._history,
            user_message=message,
            system_prompt=self._system_prompt,
            mode=self._mode,
            subagent_providers=self._subagent_providers,
            subagent_mcp_allow=self._subagent_mcp_allow,
            subagent_skills=self._subagent_skills,
        )

        agent, _tools = await self._factory.build(self.run_config)
        input_messages = self._build_input_messages(self._history, message)
        graph_input = self._build_graph_input(input_messages, message, self._mode)

        translator = EventTranslator(self._conversation_id, self._turn_id)

        try:
            async for chunk in agent.astream(
                graph_input,
                stream_mode=["messages", "updates", "custom"],
                subgraphs=True,
                version="v2",
                config={"recursion_limit": self.run_config.max_steps},
            ):
                self._accumulate_reasoning(chunk)

                for stream_event in translator.translate(chunk):
                    self._accumulate_event(stream_event)
                    yield stream_event
        except AgentRunError:
            raise
        except Exception as exc:
            logger.error("Agent execution failed: %s", exc, exc_info=True)
            raise AgentRunError(
                str(exc), recoverable=not isinstance(exc, (KeyboardInterrupt, SystemExit))
            ) from exc

    def _accumulate_reasoning(self, chunk: Any) -> None:
        """Extract reasoning_content from chunk in v2 subgraph format.

        chunk is a dict: {"type": "messages"|"updates", "ns": (...), "data": ...}
        For type="messages", data is (msg, metadata).
        """
        if not isinstance(chunk, dict):
            return
        if chunk.get("type") != "messages":
            return
        data = chunk.get("data")
        if not isinstance(data, tuple) or len(data) < 2:
            return
        msg = data[0]
        if msg is None:
            return
        reasoning_content = getattr(msg, "reasoning_content", None)
        if not reasoning_content and hasattr(msg, "additional_kwargs"):
            reasoning_content = msg.additional_kwargs.get("reasoning_content")
        if reasoning_content and isinstance(reasoning_content, str):
            self.reasoning_content += reasoning_content

    def _accumulate_event(self, stream_event: StreamEvent) -> None:
        if stream_event.type == StreamEventType.TOKEN:
            chunk = stream_event.payload.get("content", "")
            if isinstance(chunk, list):
                parts = []
                for block in chunk:
                    if isinstance(block, str):
                        parts.append(block)
                    elif isinstance(block, dict) and block.get("type") == "text":
                        parts.append(block.get("text", ""))
                chunk = "".join(parts)
            if chunk:
                self.assistant_content += chunk
        elif stream_event.type == StreamEventType.TOOL_START:
            self.tool_calls.append(stream_event)

    @staticmethod
    def _build_input_messages(
        history: list[ConversationMessage], user_message: str
    ) -> list[Any]:
        from langchain_core.messages import HumanMessage

        messages: list[Any] = [msg.to_langchain_message() for msg in history]
        messages.append(HumanMessage(content=user_message))
        return messages

    @staticmethod
    def _build_graph_input(
        messages: list[Any],
        user_message: str,
        mode: str,
    ) -> dict[str, Any]:
        graph_input: dict[str, Any] = {"messages": messages}
        if mode in ("audit", "audit-slim"):
            graph_input["audit_task"] = user_message
            graph_input["stage_outputs"] = {}
            graph_input["gapfill_count"] = 0
            graph_input["feedback_count"] = 0
            graph_input["current_stage"] = ""
        return graph_input

File: src/engine/agentRunner.ts
/**
 * Agent construction and streaming execution for Ruflo and audit turns.
 *
 * For audit-mode runs, uses LangGraph subgraph streaming (`subgraphs: true`)
 * so that every token / tool call / result is attributable to either the
 * orchestrator or a named subagent. The EventTranslator resolves namespace
 * tuples into subagent names and emits StreamEvents with a `subagent` field.
 *
 * For Ruflo, uses a primary LangGraph ReAct agent with controlled delegation.
 */

import { HumanMessage } from "@langchain/core/messages";
import type { BaseMessage } from "@langchain/core/messages";

import { AgentFactory } from "@/engine/agentFactory";
import type { McpServerDTO, ModelProviderDTO } from "@/engine/dto";
import { AgentRunError } from "@/engine/errors";
import { McpClientPool } from "@/engine/mcpPool";
import {
  type AgentRunConfig,
  ConversationMessage,
  makeAgentRunConfig,
  ResolvedSkill,
  StreamEvent,
} from "@/engine/models";
import { RagConfig } from "@/config";
import { EventTranslator } from "@/engine/streaming";
import { StreamEventType } from "@/engine/enums";

type MaybePromise<T> = T | Promise<T>;
type StreamableAgent = {
  stream?: (...args: unknown[]) => MaybePromise<AsyncIterable<unknown>>;
};

export interface AgentRunnerOptions {
  factory: AgentFactory;
  pool: McpClientPool;
  provider: ModelProviderDTO;
  enabledServers: McpServerDTO[];
  skill: ResolvedSkill | null;
  history: ConversationMessage[];
  systemPrompt: string;
  conversationId: string;
  turnId: string;
  mode?: string;
  subagentProviders?: Record<string, ModelProviderDTO>;
  subagentMcpAllow?: Record<string, string[] | null>;
  subagentSkills?: Record<string, ResolvedSkill | null>;
  rag?: RagConfig | null;
}

/** Build a LangGraph agent, run it with streaming, and yield events. */
export class AgentRunner {
  assistantContent = "";
  reasoningContent = "";
  toolCalls: StreamEvent[] = [];
  runConfig: AgentRunConfig | null = null;
  /** Latest `stage_outputs` seen on the outer graph stream (audit modes). */
  private stageOutputs: Record<string, string> = {};

  /**
   * The content to persist/display for this turn.
   *
   * For audit / audit-slim the orchestrator emits no tokens of its own; each
   * stage's text lands in `stage_outputs`. The user-facing answer is the
   * `report` stage (falling back to the last stage that produced output), not
   * the concatenation of every stage's streamed tokens. For Ruflo there are no
   * stage outputs, so this is just the accumulated assistant tokens.
   */
  get finalContent(): string {
    const report = this.stageOutputs.report;
    if (typeof report === "string" && report.trim()) return report;
    const stages = Object.values(this.stageOutputs).filter((v) => v && v.trim());
    if (stages.length) return stages[stages.length - 1]!;
    return this.assistantContent;
  }

  private readonly opts: Required<
    Omit<AgentRunnerOptions, "rag">
  > & { rag: RagConfig };

  constructor(options: AgentRunnerOptions) {
    this.opts = {
      factory: options.factory,
      pool: options.pool,
      provider: options.provider,
      enabledServers: options.enabledServers,
      skill: options.skill,
      history: options.history,
      systemPrompt: options.systemPrompt,
      conversationId: options.conversationId,
      turnId: options.turnId,
      mode: options.mode ?? "audit",
      subagentProviders: options.subagentProviders ?? {},
      subagentMcpAllow: options.subagentMcpAllow ?? {},
      subagentSkills: options.subagentSkills ?? {},
      rag: options.rag ?? new RagConfig(),
    };
  }

  async *run(message: string): AsyncIterableIterator<StreamEvent> {
    const o = this.opts;
    this.runConfig = makeAgentRunConfig({
      conversationId: o.conversationId,
      provider: o.provider,
      skill: o.skill,
      enabledServers: o.enabledServers,
      messageHistory: o.history,
      userMessage: message,
      systemPrompt: o.systemPrompt,
      mode: o.mode,
      subagentProviders: o.subagentProviders,
      subagentMcpAllow: o.subagentMcpAllow,
      subagentSkills: o.subagentSkills,
      rag: o.rag,
    });

    const [agent] = await o.factory.build(this.runConfig);
    const inputMessages = AgentRunner.buildInputMessages(o.history, message);
    const graphInput = AgentRunner.buildGraphInput(inputMessages, message, o.mode);

    const translator = new EventTranslator(o.conversationId, o.turnId);

    const streamFn = (agent as StreamableAgent).stream;
    if (typeof streamFn !== "function") {
      throw new AgentRunError("Compiled agent does not expose a stream() method.", true);
    }

    let stream: AsyncIterable<unknown>;
    try {
      stream = await streamFn.call(agent, graphInput, {
        streamMode: ["messages", "updates", "custom"],
        subgraphs: true,
        recursionLimit: this.runConfig.maxSteps,
        configurable: { thread_id: o.conversationId },
      });
      if (!stream || typeof stream[Symbol.asyncIterator] !== "function") {
        throw new Error("Compiled agent stream() did not return an AsyncIterable.");
      }
    } catch (exc) {
      throw new AgentRunError(String(exc instanceof Error ? exc.message : exc), true);
    }

    try {
      for await (const chunk of stream) {
        this.accumulateReasoning(chunk);
        this.captureStageOutputs(chunk);
        for (const streamEvent of translator.translate(chunk)) {
          this.accumulateEvent(streamEvent);
          yield streamEvent;
        }
      }
    } catch (exc) {
      if (exc instanceof AgentRunError) throw exc;
      throw new AgentRunError(String(exc instanceof Error ? exc.message : exc), true);
    }
  }

  /**
   * Capture `stage_outputs` from outer-namespace update chunks (audit modes).
   * The outer stream emits `[[], "updates", { <node>: { stage_outputs } }]` as
   * each subagent node returns; we keep the latest merged view.
   */
  private captureStageOutputs(chunk: unknown): void {
    if (!Array.isArray(chunk)) return;
    const ns = chunk.length === 3 ? chunk[0] : [];
    const mode = chunk.length === 3 ? chunk[1] : chunk.length === 2 ? chunk[0] : null;
    const data = chunk.length === 3 ? chunk[2] : chunk.length === 2 ? chunk[1] : null;
    // Only outer-namespace updates carry the merged node return values.
    if (mode !== "updates" || (Array.isArray(ns) && ns.length !== 0)) return;
    if (!data || typeof data !== "object") return;
    for (const delta of Object.values(data as Record<string, unknown>)) {
      const so =
        delta && typeof delta === "object"
          ? (delta as { stage_outputs?: unknown }).stage_outputs
          : undefined;
      if (so && typeof so === "object") {
        for (const [k, v] of Object.entries(so as Record<string, unknown>)) {
          if (typeof v === "string") this.stageOutputs[k] = v;
        }
      }
    }
  }

  private accumulateReasoning(chunk: unknown): void {
    if (!Array.isArray(chunk)) return;
    // [ns, mode, data] (subgraphs on) or [mode, data].
    const mode = chunk.length === 3 ? chunk[1] : chunk.length === 2 ? chunk[0] : null;
    const data = chunk.length === 3 ? chunk[2] : chunk.length === 2 ? chunk[1] : null;
    if (mode !== "messages") return;
    if (!Array.isArray(data) || data.length < 2) return;
    const msg = data[0] as
      | { reasoning_content?: unknown; additional_kwargs?: Record<string, unknown> }
      | null
      | undefined;
    if (!msg) return;
    let reasoning = msg.reasoning_content;
    if (!reasoning && msg.additional_kwargs) {
      reasoning = msg.additional_kwargs.reasoning_content;
    }
    if (typeof reasoning === "string" && reasoning) {
      this.reasoningContent += reasoning;
    }
  }

  private accumulateEvent(streamEvent: StreamEvent): void {
    if (streamEvent.type === StreamEventType.TOKEN) {
      let chunk = streamEvent.payload.content as unknown;
      if (Array.isArray(chunk)) {
        const parts: string[] = [];
        for (const block of chunk) {
          if (typeof block === "string") parts.push(block);
          else if (block && typeof block === "object" && (block as { type?: string }).type === "text") {
            parts.push(String((block as { text?: string }).text ?? ""));
          }
        }
        chunk = parts.join("");
      }
      if (typeof chunk === "string" && chunk) this.assistantContent += chunk;
    } else if (streamEvent.type === StreamEventType.TOOL_START) {
      this.toolCalls.push(streamEvent);
    }
  }

  private static buildInputMessages(
    history: ConversationMessage[],
    userMessage: string,
  ): BaseMessage[] {
    const messages: BaseMessage[] = history.map((m) => m.toLangchainMessage());
    messages.push(new HumanMessage({ content: userMessage }));
    return messages;
  }

  private static buildGraphInput(
    messages: BaseMessage[],
    userMessage: string,
    mode: string,
  ): Record<string, unknown> {
    const graphInput: Record<string, unknown> = { messages };
    if (mode === "audit" || mode === "audit-slim") {
      graphInput.audit_task = userMessage;
      graphInput.stage_outputs = {};
      graphInput.gapfill_count = 0;
      graphInput.feedback_count = 0;
      graphInput.current_stage = "";
    }
    return graphInput;
  }
}
